{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "This is a custom viewer for Hi-C data. I don't know why I did this when other, better viewers are available for free, but what the hell. Currently, it's quite janky and uses a fixed field of view at four zoom rates. It inter-operates with some confusing Python code. I should probably clean all this up...\n",
    "\n",
    "Desired functionality:\n",
    "\n",
    "(1) Show Hi-C data at multiple resolutions\n",
    "(2) Use multiple colormaps\n",
    "(3) Allow on-the-fly contrast adjustment\n",
    "(4) Allow positions to be clicked and stored\n",
    "(5) Show genomic data alongside Hi-C maps [this could be greatly improved]\n",
    "\n",
    "Improvements:\n",
    "\n",
    "-Make it so it doesn't demand mel chromosomes<br>\n",
    "-Improve stability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Set folders and files:\n",
    "    folder = folder containing viewer data files\n",
    "    save_folder = folder to save viewer snapshots to\n",
    "    track_folder = folder containing genomic data tracks\n",
    "'''\n",
    "folder = '/Users/MStadler/Bioinformatics/Projects/HiC/data_2018/viewer_data/HiC-53_54_multi'\n",
    "save_folder = '/Users/MStadler/Bioinformatics/Projects/HiC/data_2018/visual_inspections/HiC-53_54'\n",
    "track_folder = '/Users/MStadler/Bioinformatics/Projects/HiC/data_2018/viewer_data/viewer_tracks/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "\n",
    "##mpl.use('TkAgg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.widgets import Slider, Button, RadioButtons, Cursor, TextBox, MultiCursor\n",
    "import tkinter as tk\n",
    "from tkinter import filedialog\n",
    "import re\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "import PyQt5\n",
    "%matplotlib qt\n",
    "\n",
    "#bug: if I start with inline then change it to qt it's fine...otherwise error if starting qt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Function to save the list of clicked coordinates to a file \n",
    "def save_coord_list(filename_coords, coord_list):\n",
    "   with open(filename_coords, 'w') as file_coords: \n",
    "        for i in range(0,len(coord_list)):\n",
    "            chrom = coord_list[i][0]\n",
    "            pos1 = str(int((coord_list[i][1] * 1000000) / 500) * 500)\n",
    "            pos2 = str(int((coord_list[i][2] * 1000000) / 500) * 500)\n",
    "            file_coords.write(chrom + ':' + pos1 + '-' + pos2 + '\\t' + chrom + '\\t' + pos1 + '\\t' + pos2 + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "''' \n",
    "Function to load Hi-C reduced matrix, flatten the top and bottom 1% of the data, takes log, \n",
    "scales to 1-1000 to account for negative numbers \n",
    "'''\n",
    "def load_process_hic_matrix(infilename):\n",
    "    x = np.genfromtxt(infilename, delimiter='\\t', skip_header=1)\n",
    "    x = x[:,1:] #take off first row which is generally bin names\n",
    "    # Next two lines take care of 0's and nan's by setting to dummy value (typically 0.1)\n",
    "    dummy_val = 0.1\n",
    "    upper_trim = 99\n",
    "    x[x == 0] = dummy_val\n",
    "    x[np.isnan(x)] = dummy_val\n",
    "    top = np.percentile(x, upper_trim)\n",
    "    x[x > top] = top\n",
    "    x = np.log(x)\n",
    "    x = (x - np.min(x)) / (np.max(x) - np.min(x)) * 1000 # scales to 1-1000 betwixt min and max values.\n",
    "    return(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "452d8ccbe72e42558ac60883eae3f177",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dropdown(description='Track:', options=('S5_DNase', 'S5_RNAP2', 'BEAF-32 0-12h', 'dCTCF 0-12h', 'CP190 0-12h',…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "NameError",
     "evalue": "name 'Map' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-d122099d7464>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    275\u001b[0m \u001b[0;31m#Draws a crosshair cursor for visually aligning Hi-C features with genomic features\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    276\u001b[0m \u001b[0mcursor\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCursor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolor\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"black\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 277\u001b[0;31m \u001b[0mm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mMap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    278\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    279\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mhandle_interaction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'Map' is not defined"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Main function, calls viewer\n",
    "'''\n",
    "#for 500 bp: 1000 bins 250,000 bp step\n",
    "\n",
    "#######################################################################\n",
    "# Core functions and classes, nuts and bolts of the viewer\n",
    "#######################################################################\n",
    "\n",
    "#Position class stores the information about the current state of the system (position, colormap, step size, etc.)\n",
    "class Position:\n",
    "    def __init__(self):\n",
    "        Position.chr = \"2R\"\n",
    "        Position.binsize = 500\n",
    "        Position.L = 9000000\n",
    "        Position.cmap = []\n",
    "        Position.step_size = 100000\n",
    "        Position.window_size = 200000\n",
    "        Position.track_data = {}\n",
    "\n",
    "#######################################################################\n",
    "#linearly combine two colormaps—required a bit of work\n",
    "\n",
    "def combine_colormaps(cm1, cm2):\n",
    "    colors1 = cm1(np.linspace(0., 1, 128))\n",
    "    colors2 = cm2(np.linspace(0., 1, 128))\n",
    "    colors = np.vstack((colors1, colors2))\n",
    "    return(mpl.colors.LinearSegmentedColormap.from_list('my_colormap', colors))\n",
    "\n",
    "#######################################################################\n",
    "#general function to update genome location when position is changed\n",
    "\n",
    "def load_chr_pos():\n",
    "    posL = pos.L\n",
    "    posR = posL + pos.window_size\n",
    "    new_file = os.path.join(folder, '_'.join([pos.chr, str(posL), str(posR)]) + '_' + str(pos.binsize) + '.txt')\n",
    "    if (os.path.isfile(new_file) or os.path.isfile(new_file + '.gz')): \n",
    "        new_img = load_process_hic_matrix(new_file)\n",
    "        img.set_data(new_img)\n",
    "        L_Mb = float(posL) / 1e6\n",
    "        R_Mb = float(posR) / 1e6\n",
    "        img.set_extent([L_Mb, R_Mb, R_Mb, L_Mb])\n",
    "        tloc.set_val(str(L_Mb))\n",
    "        ax.set_title(pos.chr + ': ' + str(L_Mb) + '-' + str(R_Mb) + ' Mb', fontsize=14)\n",
    "        img.set_cmap(pos.cmap)\n",
    "        fig.canvas.draw_idle()\n",
    "    else:\n",
    "        ax.set_title('Position is out of bounds!', fontsize=14)\n",
    "        \n",
    "    #update WIG track data\n",
    "    track_binsize = 500\n",
    "    track_binL = int(posL / track_binsize)\n",
    "    track_binR = int(track_binL + int(pos.window_size / track_binsize))\n",
    "    ax_track.cla()\n",
    "    ax_track.plot(pos.track_data[pos.chr][range(track_binR, track_binL, -1)], range(track_binL, track_binR))\n",
    "    ax_track.set_ylim(track_binL, track_binR)\n",
    "\n",
    "#######################################################################\n",
    "#function for loading a new genomic dataset ('track')\n",
    "\n",
    "def load_track_data(track):\n",
    "    if (track == 'S5_DNase'): \n",
    "        track_file = 'S5_DNase_rep1_4col_dm6_binned500.txt'\n",
    "    if (track == 'S5_RNAP2'):\n",
    "        track_file = 'S5_polII_FDR25_4col_dm6_binned500.txt'\n",
    "    if (track == 'BEAF-32 0-12h'):\n",
    "        track_file = 'BEAF-32_0_12h_4col_dm6_binned500.txt'\n",
    "    if (track == 'dCTCF 0-12h'):\n",
    "        track_file = 'dCTCF_0_12h_4col_dm6_binned500.txt'\n",
    "    if (track == 'CP190 0-12h'):\n",
    "        track_file = 'CP190_0_12h_4col_dm6_binned500.txt'\n",
    "    if (track == 'promoters'):\n",
    "        track_file = 'dm6_promoters.txt'\n",
    "    track_file_path = os.path.join(track_folder, track_file)\n",
    "    max_chr_size = 4e7 #maximum allowable size of chromosomes for allocating arrays\n",
    "    track_binsize = 500 #binsize of the track data\n",
    "    pos.track_data = {}\n",
    "    with open(track_file_path, 'r') as infile:\n",
    "        for line in infile:\n",
    "            items = line.split()\n",
    "            (chr, posL, posR, val) = items\n",
    "            chr = re.sub('chr', '', chr)\n",
    "            if (chr not in pos.track_data):\n",
    "                pos.track_data[chr] = np.zeros(int(max_chr_size / track_binsize))\n",
    "            track_bin = int(int(posL) / track_binsize)\n",
    "            pos.track_data[chr][track_bin] = float(val)    \n",
    "\n",
    "#######################################################################\n",
    "#Set up initial variables, containers, axes etc.\n",
    "#######################################################################\n",
    "\n",
    "pos = Position()\n",
    "pos.cmap = combine_colormaps(plt.cm.Blues_r, plt.cm.YlOrBr)\n",
    "axcolor=\"white\"\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "fig.tight_layout()\n",
    "plt.subplots_adjust(bottom=0.20, top=0.95)\n",
    "folder_name_m = re.search(r'[^/]+$',folder) #matches everything after last slash\n",
    "folder_name = folder_name_m.group()\n",
    "fig.text(0.05, 0.05, folder_name, fontsize=20)\n",
    "img = plt.imshow(np.zeros((100,100)), vmin=0, vmax=1000, cmap=pos.cmap, interpolation=\"none\")\n",
    "ax_track = plt.axes([0.8, 0.2, 0.1, 0.75],) #axes for WIG track data\n",
    "\n",
    "#######################################################################\n",
    "#Functions for the interactive features of the viewer\n",
    "#######################################################################\n",
    "#######################################################################\n",
    "# Press any key to store the cursor position in coords\n",
    "\n",
    "coords = []\n",
    "def onkeypress(event):\n",
    "    if(event.inaxes == ax):\n",
    "        global coords\n",
    "        coords.append((pos.chr, event.xdata, event.ydata))\n",
    "        print((pos.chr, event.xdata, event.ydata))\n",
    "\n",
    "cid = fig.canvas.mpl_connect('key_press_event', onkeypress)\n",
    "\n",
    "# Click mouse to display position on track\n",
    "def onclick(event):\n",
    "    if(event.inaxes == ax):\n",
    "        plt.sca(ax_track) #set to active axis\n",
    "       # print(plt.ylim()[0])\n",
    "       # print(plt.ylim()[1])\n",
    "        plt.hlines(plt.ylim()[1] - (event.ydata * 1e6 / Position.binsize) + plt.ylim()[0] + 1, plt.xlim()[0]*0.5,plt.xlim()[1] * 0.5)\n",
    "cid = fig.canvas.mpl_connect('button_press_event', onclick)\n",
    "\n",
    "def onrelease(event):\n",
    "    if(event.inaxes == ax):\n",
    "        load_chr_pos()\n",
    "cid = fig.canvas.mpl_connect('button_release_event', onrelease)\n",
    "\n",
    "#######################################################################\n",
    "#Radio buttons to change colormap\n",
    "\n",
    "ax_colors = plt.axes([0.05, 0.3, 0.10, 0.15])\n",
    "colors = RadioButtons(ax_colors, ('BlOr', 'Rd', 'RdGy'), active=0)\n",
    "\n",
    "def update_colors(selection):\n",
    "    if (selection == 'BlOr'):\n",
    "        colormap = combine_colormaps(plt.cm.Blues_r, plt.cm.YlOrBr)\n",
    "    elif (selection == 'Rd'):\n",
    "        colormap = plt.cm.Reds\n",
    "    elif (selection == 'RdGy'):\n",
    "        colormap = plt.cm.RdGy_r\n",
    "    pos.cmap = colormap\n",
    "    img.set_cmap(pos.cmap)\n",
    "    fig.canvas.draw_idle()   \n",
    "colors.on_clicked(update_colors)\n",
    "\n",
    "#######################################################################        \n",
    "#Radio buttons to change chromosome\n",
    "\n",
    "ax_chrom = plt.axes([0.05, 0.5, 0.05, 0.15])\n",
    "chrom = RadioButtons(ax_chrom, ('2L', '2R', '3L', '3R', 'X'), active=0)\n",
    "\n",
    "def update_chromosome(selection):\n",
    "    pos.chr = selection\n",
    "    load_chr_pos()\n",
    "chrom.on_clicked(update_chromosome)\n",
    "\n",
    "#######################################################################\n",
    "#Radio buttons to change bin size (resolution)\n",
    "\n",
    "ax_binsize = plt.axes([0.05, 0.8, 0.05, 0.15])\n",
    "binsize = RadioButtons(ax_binsize, ('500', '1000', '2000', '4000'), active=0)\n",
    "\n",
    "def update_binsize(selection):\n",
    "    pos.binsize = int(selection)\n",
    "    pos.step_size = int(selection) * 200\n",
    "    pos.window_size = int(selection) * 400\n",
    "    pos.L = int(pos.L / pos.step_size) * pos.step_size #finds nearest position that is a valid start position for this bin size\n",
    "    load_chr_pos()\n",
    "binsize.on_clicked(update_binsize)\n",
    "\n",
    "#######################################################################\n",
    "#Clickable button to save current viewer image as png into save_folder designated at top\n",
    "\n",
    "ax_savefig = plt.axes([0.05, 0.7, 0.07, 0.05]) #left, bottom, width, height\n",
    "savefig = Button(ax_savefig, 'Save', color=\"orange\")\n",
    "\n",
    "def execute_savefig(selection):\n",
    "    file_path = os.path.join(save_folder, (pos.chr + '_' + str(pos.L) + '.png'))\n",
    "    fig_saving, ax_saving = plt.subplots()\n",
    "    plt.imshow(img.get_array(), vmin=scontrastlower.val, vmax=scontrastupper.val, cmap = pos.cmap, interpolation=\"none\", extent = img.get_extent())\n",
    "    mpl.pyplot.savefig(file_path)\n",
    "    mpl.pyplot.close()\n",
    "savefig.on_clicked(execute_savefig)\n",
    "\n",
    "#######################################################################\n",
    "#Editable textbox storing genomic position (displays and receives values)\n",
    "\n",
    "ax_position_text = plt.axes([0.50, 0.03, 0.05, 0.05], facecolor=axcolor)\n",
    "tloc = TextBox(ax_position_text, 'Mb')\n",
    "\n",
    "def location_text_callback(text):\n",
    "    pos_entered = int(float(text)* 1000000) #position in megabases\n",
    "    if (pos_entered % pos.step_size == 0):\n",
    "        pos.L = pos_entered\n",
    "        load_chr_pos()\n",
    "tloc.on_submit(location_text_callback)\n",
    "\n",
    "#######################################################################\n",
    "#contrast adjustment sliders\n",
    "\n",
    "ax_contrast_lower = plt.axes([0.25, 0.1, 0.5, 0.02], facecolor=axcolor)\n",
    "ax_contrast_upper = plt.axes([0.25, 0.13, 0.5, 0.02], facecolor=axcolor)\n",
    "scontrastlower = Slider(ax_contrast_lower, 'Lower', 0, 1000, valinit=0, valstep=1, valfmt=\"%1.0f\", fill=False, edgecolor=\"blue\", linewidth=1)\n",
    "scontrastupper = Slider(ax_contrast_upper, 'Upper', 0, 1000, valinit=1000, valstep=1, valfmt=\"%1.0f\", fill=False, edgecolor=\"orange\", linewidth=1)\n",
    "\n",
    "def update_contrast(val):\n",
    "    lower = int(scontrastlower.val)\n",
    "    img.set_norm(mpl.colors.Normalize(vmin=scontrastlower.val, vmax=scontrastupper.val))\n",
    "    fig.canvas.draw_idle()\n",
    "scontrastlower.on_changed(update_contrast)\n",
    "scontrastupper.on_changed(update_contrast)\n",
    "\n",
    "#######################################################################\n",
    "#left and right scroll buttons\n",
    "\n",
    "ax_200kb_L = plt.axes([0.40, 0.03, 0.05, 0.05])\n",
    "ax_200kb_R = plt.axes([0.58, 0.03, 0.05, 0.05])\n",
    "ax_1Mb_L = plt.axes([0.30, 0.03, 0.05, 0.05])\n",
    "ax_1Mb_R = plt.axes([0.68, 0.03, 0.05, 0.05])\n",
    "\n",
    "b200L = Button(ax_200kb_L, '<', color=\"dodgerblue\", hovercolor=\"darkorange\")\n",
    "b200R = Button(ax_200kb_R, '>', color=\"dodgerblue\", hovercolor=\"darkorange\")\n",
    "b1MbL = Button(ax_1Mb_L, '<<', color=\"dodgerblue\", hovercolor=\"darkorange\")\n",
    "b1MbR = Button(ax_1Mb_R, '>>', color=\"dodgerblue\", hovercolor=\"darkorange\")\n",
    "\n",
    "def update_b200L(click):\n",
    "    pos.L = int(pos.L - pos.step_size)\n",
    "    load_chr_pos()\n",
    "b200L.on_clicked(update_b200L)\n",
    "\n",
    "def update_b200R(click):\n",
    "    pos.L = int(pos.L + pos.step_size)\n",
    "    load_chr_pos()\n",
    "b200R.on_clicked(update_b200R)\n",
    "\n",
    "def update_b1MbL(click):\n",
    "    pos.L = int(pos.L - (pos.step_size * 5))\n",
    "    load_chr_pos()\n",
    "b1MbL.on_clicked(update_b1MbL)\n",
    "\n",
    "def update_b1MbR(click):\n",
    "    pos.L = int(pos.L + (pos.step_size * 5))\n",
    "    load_chr_pos()\n",
    "b1MbR.on_clicked(update_b1MbR)\n",
    "\n",
    "#######################################################################\n",
    "# Dropdown menu to display track data\n",
    "\n",
    "load_track_data('S5_DNase')\n",
    "load_chr_pos()\n",
    "\n",
    "def track_selector_callback(change):\n",
    "    if change['type'] == 'change' and change['name'] == 'value':\n",
    "        load_track_data(track_selector.value)\n",
    "        load_chr_pos()\n",
    "    \n",
    "track_selector = widgets.Dropdown(\n",
    "    options=['S5_DNase', 'S5_RNAP2', 'BEAF-32 0-12h', 'dCTCF 0-12h', 'CP190 0-12h', 'promoters'],\n",
    "    value='S5_DNase',\n",
    "    description='Track:',\n",
    "    disabled=False,\n",
    ")\n",
    "track_selector.observe(track_selector_callback)\n",
    "display(track_selector)\n",
    "\n",
    "#######################################################################\n",
    "#Multicursor\n",
    "#multi = MultiCursor(fig.canvas, (ax_track, ax), color='r', lw=1, horizOn=True, vertOn=True)\n",
    "\n",
    "#Draws a crosshair cursor for visually aligning Hi-C features with genomic features\n",
    "cursor = Cursor(ax, color=\"black\")\n",
    "m = Map()\n",
    "\n",
    "def handle_interaction(**kwargs):\n",
    "    if kwargs.get('type') == 'mousemove':\n",
    "        pass\n",
    "\n",
    "m.on_interaction(handle_interaction)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "save_coord_list('/Users/MStadler/Bioinformatics/Projects/HiC/data_2018/visual_inspections/20190619_HiC-59/20190702_1000bp_58_59_visualDiffs.txt',coords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('2R', 9.077297035899186, 9.074326676907322),\n",
       " ('2R', 9.024455254025147, 9.032544802867383)]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda env:hic3]",
   "language": "python",
   "name": "conda-env-hic3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
